### ISSUE TITLE
proxyForInterface should make it possible to specify the attribute name that the wrapped object is assigned to

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2895]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#2895|
|Type|enhancement|
|Created|2007-11-13 21:12:16Z|
|Branch|https://github.com/twisted/twisted/tree/pfi-custom-attribute-2895-4|

```
class Foo(proxyForInterface(IFoo, originalAttribute="result")):
    def blah(self):
        return self.result.blah() - 1
```

would be cool.

<details><summary>Searchable metadata</summary>

```
trac-id__2895 2895
type__enhancement enhancement
reporter__radix radix
priority__highest highest
milestone__ 
branch__branches_pfi_custom_attribute_2895_4 branches/pfi-custom-attribute-2895-4
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1194988336000000 1194988336000000
changetime__1199912677000000 1199912677000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

This should be fixed before release so that we aren't lugging around another deprecated API.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [21965]) Branching to 'pfi-custom-attribute-2895'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [21966]) Add the new attribute, and test the customization.

Refs [#2895](https://github.com/twisted/twisted/issues/2895)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

That was easy :). Note that it's done without breaking anything, so first comment about deprecated API doesn't stand.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @therve|
|-|-|

Waiting for [#2901](https://github.com/twisted/twisted/issues/2901)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21995]) Branching to 'pfi-custom-attribute-2895-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [21996]) Do it again.

Refs [#2895](https://github.com/twisted/twisted/issues/2895)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [21998]) Doc.

Refs [#2895](https://github.com/twisted/twisted/issues/2895)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [22139]) Branching to 'pfi-custom-attribute-2895-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve removed owner|
|-|-|

Back to back to back to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

Looks good.  The only suggestion I have is that we may want to change the existing uses of ```proxyForInterface``` in Twisted to pass a sensible value for this new parameter.

Maybe see what jml thinks about that, since the existing uses are in trial.

Generally I think Twisted itself should never just use _original_, since the choice of name here is a free opportunity to make code easier to read.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

Do you think we should do it there or open another ticket?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

The branch isn't big now and adding this won't add much more, and fixing the uses in Twisted is the real motivation to make this change, so I think it's fine to do it here.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

(In [22279]) Branching to 'pfi-custom-attribute-2895-4'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve removed owner|
|-|-|

Ready to review!

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @radix<br>@radix set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_20"></a>|@radix set owner to @therve<br>@radix set status to `new`|
|-|-|


 - Why B{}? I think C{} is more appropriate.

 - Also, given that the usages of proxyForInterface you changed didn't document their original attribute in the first place, it seems like it might be a good idea to give these original attributes names that start with _.

Other than that I'm happy with the branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set status to `closed`|
|-|-|

(In [22304]) Merge pfi-custom-attribute-2895-4

Author: therve
Reviewers: radix, exarkun
Fixes [#2895](https://github.com/twisted/twisted/issues/2895)

Add the possibility to proxyForInterface to specify the attribute where the
proxied object is stored.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
